linkerd/app/outbound/src/ingress.rs

```diff
@@ -252,69 +259,132 @@ impl Param<profiles::LookupAddr> for Http<RequestTarget> {
     }
 }
 
-impl svc::Param<http::Logical> for Http<http::Logical> {
-    fn param(&self) -> http::Logical {
-        self.parent.clone()
+impl svc::Param<http::LogicalAddr> for Http<Logical> {
+    fn param(&self) -> http::LogicalAddr {
+        http::LogicalAddr(self.parent.addr.clone())
     }
 }
 
-impl svc::Param<http::normalize_uri::DefaultAuthority> for Http<http::Logical> {
+impl svc::Param<http::normalize_uri::DefaultAuthority> for Http<Logical> {
     fn param(&self) -> http::normalize_uri::DefaultAuthority {
-        http::normalize_uri::DefaultAuthority(match &self.parent {
-            http::Logical::Route(addr, _) => Some(addr.as_http_authority()),
-            http::Logical::Forward(..) => None,
-        })
+        http::normalize_uri::DefaultAuthority(Some(self.parent.addr.to_http_authority()))
     }
 }
 
-impl TryFrom<discover::Discovery<Http<RequestTarget>>> for Http<http::Logical> {
+impl svc::Param<watch::Receiver<http::Routes>> for Http<Logical> {
+    fn param(&self) -> watch::Receiver<http::Routes> {
+        self.parent.routes.clone()
+    }
+}
+
+impl TryFrom<discover::Discovery<Http<RequestTarget>>> for Http<Logical> {
     type Error = ProfileRequired;
 
     fn try_from(
         parent: discover::Discovery<Http<RequestTarget>>,
     ) -> std::result::Result<Self, Self::Error> {
         match (
-            &**parent,
-            svc::Param::<Option<profiles::Receiver>>::param(&parent),
+            (**parent).clone(),
+            svc::Param::<Option<profiles::Receiver>>::param(&parent).map(watch::Receiver::from),
         ) {
-            (RequestTarget::Named(addr), Some(profile)) => {
-                if let Some(profiles::LogicalAddr(addr)) = profile.logical_addr() {
-                    return Ok(Http {
-                        version: (*parent).param(),
-                        parent: http::Logical::Route(addr, profile),
-                    });
-                }
-
-                let (addr, metadata) = profile
-                    .endpoint()
-                    .ok_or_else(|| ProfileRequired(addr.clone()))?;
+            (RequestTarget::Named(addr), profile) => {
+                let profile = profile.ok_or_else(|| ProfileRequired(addr.clone()))?;
+                let mut route =
+                    mk_route(&*profile.borrow()).ok_or_else(|| ProfileRequired(addr.clone()))?;
+                let routes = http::spawn_routes(profile, move |p: &profiles::Profile| {
+                    if let Some(r) = mk_route(p) {
+                        route = r;
+                    }
+                    route.clone()
```

Comment:
would it make sense for the closure passed to `spawn_routes` to return an `Option`, and only update the watch if the `Option` is `Some`? this looks like we will spuriously update the watch with the existing state if `mk_route` returns `None`. i don't think this is a major issue, but it's not strictly necessary...